runtime/interpreter/big.go

```diff
@@ -140,3 +141,31 @@ func BigEndianBytesToSignedBigInt(b []byte) *big.Int {
 func BigEndianBytesToUnsignedBigInt(b []byte) *big.Int {
 	return new(big.Int).SetBytes(b)
 }
+
+func BigIntSqrt(interpreter *Interpreter, value *big.Int, locationRange LocationRange) UFix64Value {
+	if value.Sign() < 0 {
+		panic(UnderflowError{
+			LocationRange: locationRange,
+		})
+	}
+
+	valueFloat := new(big.Float).SetPrec(256).SetInt(value)
+	res := new(big.Float).SetPrec(256).SetMode(big.ToZero).Sqrt(valueFloat)
+
+	valueGetter := func() uint64 {
+		res.Mul(res, new(big.Float).SetPrec(256).SetInt(sema.Fix64FactorBig))
+
+		resInt := new(big.Int)
+		res.Int(resInt)
+
+		if !resInt.IsUint64() {
+			panic(OverflowError{
+				LocationRange: locationRange,
+			})
+		}
```

Comment:
Referring to the issue https://github.com/onflow/cadence/issues/1151
> ln(x), log(x,b), pow(x,y), sqrt(x), exp(x)

In my opinion, these functions are most useful if their output type is a floating point number. This is the overwhelming amount of use cases, the inputs will not permit an integer type output. I am happy with the inputs being integers, float, etc for convenience. Though, the output could be a float. 

There are cases, where the developer knows that the input `x` for `sqrt(x)` satisfies  `x = y^2` for some non-negative integer `y`.  But this is rarely that case, and developers could use functions like `round`, `floor`, `ceiling` to convert the float to an integer.

Reply:
Cadence does not have floating point numbers, only [fixed-point numbers](https://cadence-lang.org/docs/language/values-and-types#fixed-point-numbers). Like for floating point numbers, the precision/scale will only make certain values representable. Over/underflow is automatically checked and leads to a program abort.

Reply:
> In my opinion, these functions are most useful if their output type is a floating point number. This is the overwhelming amount of use cases, the inputs will not permit an integer type output. I am happy with the inputs being integers, float, etc for convenience. Though, the output could be a float.

@AlexHentschel I think there is some confusion. The output of the `Sqrt` function is always a fixed-point number (not an integer) which is the closest type we have in Cadence to a floating point number.

`Sqrt<T>` will always return `UFix64` i.e. 64-bit fixed point number.

Reply:
I think the confusion is due to the line that you have highlighted.

```
		res.Mul(res, new(big.Float).SetPrec(256).SetInt(sema.Fix64FactorBig))

		resInt := new(big.Int)
		res.Int(resInt)

		if !resInt.IsUint64() {
			panic(OverflowError{
				LocationRange: locationRange,
			})
		}
```

This snippet is not converting the result to an integer. Notice the multiplication by `Fix64FactorBig` (10^8). This is only truncating the output after 8 decimal places which is the level of precision `UFix64` supports.

Reply:
Thanks @turbolent for the clarification. I think I am lacking some necessary context here: while I interpreted `UFix64Value` as an integer value type, I realize now that it is a [Rational Number](https://en.wikipedia.org/wiki/Rational_number) representation. That is great. Sorry for my lack of context and thanks for explaining. 

Just to make sure we are on the same page:
https://github.com/onflow/cadence/blob/90b485fe2ada7dc3b752981a22a97bf9f4044d7c/runtime/interpreter/big.go#L167
A fixed-point number essentially means we have some notion of "smallest atomic unit" that cannot be split further. Anything smaller than a "smallest atomic unit" is truncated to zero. This `uint64` counts the number of "smallest atomic unit". Is that correct?

Reply:
Yes, sounds correct to me.

UFix64 represents floating point number up to a fixed precision (8 decimal places).

So if you want to represent Pi in Cadence which is `3.14159265359...`, you could only store it up to 8 decimal places in `UFix64`. So you can only represent it as:

```
let pi = UFix64.fromString("3.14159265");  // notice that I only included 8 decimal places.
```

The underlying value stored inside `UFix64` will be the value multiplied by 10^8 which is `314159265` ("smallest atomic unit" of your explanation).

cc: @turbolent

Reply:
thanks @darkdrag00nv2 for the detailed explanation. Really helpful!